.github/workflows/workflow.yml

```diff
@@ -1,38 +1,53 @@
 name: CI
 
 on:
-  push:
+  pull_request:
     branches:
       - master
-  pull_request:
+  push:
     branches:
       - master
 
 jobs:
   build:
-    runs-on: ubuntu-latest
+    strategy:
+      fail-fast: false
+      matrix:
+        os:
+          - macos-latest
+          - ubuntu-latest
+        ocaml-version:
+          - 4.10.1
+
+    runs-on: ${{ matrix.os }}
+
     steps:
       - name: Checkout code
         uses: actions/checkout@v2
 
-      - name: Cache
+      - name: Retrieve opam cache
         uses: actions/cache@v2
+        id: cache-opam
         with:
-          path: /home/runner/.opam
-          key: ez-api-opam-cache
+          path: ~/.opam
+          key: v1-${{ runner.os }}-opam-${{ matrix.ocaml-version }}-${{ hashFiles('ez-api.opam') }}
+          restore-keys: |
+            v1-${{ runner.os }}-opam-${{ matrix.ocaml-version }}-
 
-      - name: Use OCaml
+      - name: Use OCaml ${{ matrix.ocaml-version }}
         uses: avsm/setup-ocaml@v1
         with:
-          ocaml-version: 4.10.0
-
-      - name: Depext
-        run: |
-          sudo apt update
-          opam depext -y geoip ocurl tls
-
-      - name: Opam install
-        run: |
-          opam install $(opam show -f depopts: . | sed -e 's/{.*}//g' -e 's/"//g')
-          opam install .
-          opam remove .
+          ocaml-version: ${{ matrix.ocaml-version }}
+
+      - run: opam pin add ez-api.dev . --no-action
+
+      - run: opam depext ez-api --with-doc --with-test --yes
+
+      - run: opam install . --deps-only --with-doc --with-test
+        if: steps.cache-opam.outputs.cache-hit != 'true'
```

Comment:
I'm still thinking about whether we should run this step even if we retrieve the cache. What do you think?

Reply:
since the cache depends on ez-api.opam, it's probably not necessary if there is no opam update before that.
I'm not sure we need --with-doc --with-test if it's deps-only